proposals/42-human-readable-mutable-names.md

```diff
@@ -0,0 +1,235 @@
+## Purpose &amp; impact 
+
+Improve the way mutable names work in our stack.
+
+- Empower developers with reliable mechanisms for publishing updates
+  - Make IPNS useful on its own, and inside of DNSLink records
+  - Future-proof the way we do human readable names via DNS interop
+
+#### Background &amp; intent
+_Describe the desired state of the world after this project? Why does that matter?_
+<!--
+Outline the status quo, including any relevant context on the problem you’re seeing that this project should solve. Wherever possible, include pains or problems that you’ve seen users experience to help motivate why solving this problem works towards top-line objectives. 
+-->
+
+We see CID, IPNS record, and DNSLink being de-facto standard for cryptographically-verifiable content-addressing and human-readable names on the distributed web. 
+
+There are three distinct types of addresses:
+  - immutable content:  `/ipfs/{cid}`
+  - mutable content
+      - cryptographic, IPNS record signed with a specific libp2p key: `/ipns/{libp2p-key-cid}` 
+      - human-readable, [DNSLink](https://dnslink.io): `/ipns/{dns-name}`
+        (which in turn, resolves to either an immutable or mutable content path from DNS TXT record)
+
+
+**There are known gaps in the mutable story:**
+- IPNS records are only published by the original author, which severely limits the utility of our mutable names
+    - It is not possible to pin/republish `/ipns/{libp2p-key-cid}` names by other peers, so when original publisher disappears, IPNS links no longer resolve and we see "IPNS rot". 
+    - **Due to IPNS rot people simply don't use it:**
+        - Instead publishing updates via `ipfs name publish {cid}`, website operators choose to constantly update DNSLink's DNS TXT record
+            - They point at reliable immutable paths (`dnslink=/ipfs/{cid}`) (no IPNS record means it resolves fast and never break).
+            - It works around the need for running a service responsible for constant republishing of IPNS record.
+            - This is a pragmatic choice, however:
+              - adds admin related to DNS updates (need to learn DNS provider's API for automating this, which introduces soft vendor-lock-in)
+              - most likely not the best choice for security-sensitive use cases where human-readable names are not required: means DNS is used in contexts where cryptographic addressing should be used instead.
+- **OS-level DNS resolver is used for all DNS TXT record lookups.** 
+  - IPFS node is unable to control which DNS resolver is used for specific TLD, all queries go to global DNS server configured in the operating system
+    - This harms competition and slows down innovation in the decentralized naming space. For example, if user changes their DNS resolver to one that is capable of resolving all TLDs over ENS, they are unable to resolve TLDs from OpenNIC or UnstoppableDomains. 
```

Comment:
n00b questions:
1. Is there a DNS resolver that is capable of resolving all TLDs over ENS?  Is this more about something that could come in the future (per https://ccnso.icann.org/sites/default/files/field-attached/presentation-dns-blockchain-ens-24jun19-en.pdf )?
2. Is this a general issue even at the OS level that a user won't be able to have DNS that works with ENS, OpenNIC, and UnstoppableDomains together?  Can someone have a local DNS proxy that can resolve using ENS, OpenNIC, UnstoppableDomains, etc.?

Reply:
Ad 1. Yes, https://eth.link provides endpoint at `https://eth.link/dns-query`, but we will use it only for resolving `*.eth`
Ad 2. Correct, my opinion is that requiring users to run DNS proxy is not realistic. Ability to specify DoH resolver per TLD  removes the need for that, but if someone wants to run own DNS resolver, they can also do this  here – just point at localhost one.